src/configure/mod.rs

```diff
@@ -443,80 +408,105 @@ fn get_offline_autoconfig(context: &Context, param: &LoginParam) -> Option<Login
-        return Ok(());
-    }
-    if was_autoconfig {
-        bail!("autoconfig did not succeed");
+) -> Result<LoginParam> {
+    // Try to infer port from socket security.
+    if param.mail_port == 0 {
+        if 0 != param.server_flags & DC_LP_IMAP_SOCKET_SSL {
+            param.mail_port = 993
+        }
+        if 0 != param.server_flags & (DC_LP_IMAP_SOCKET_STARTTLS | DC_LP_IMAP_SOCKET_PLAIN) {
+            param.mail_port = 143
+        }
     }
 
-    progress!(context, 650 + variation * 30);
+    if param.mail_port == 0 {
+        // Neither port nor security is set.
+        //
+        // Try common secure combinations.
 
-    if manually_set_param.server_flags & DC_LP_IMAP_SOCKET_FLAGS == 0 {
-        param.server_flags &= !(DC_LP_IMAP_SOCKET_FLAGS);
-        param.server_flags |= DC_LP_IMAP_SOCKET_STARTTLS;
+        // Try TLS over port 993
+        param.server_flags &= !(DC_LP_IMAP_SOCKET_FLAGS as i32);
+        param.server_flags |= DC_LP_IMAP_SOCKET_SSL as i32;
+        param.mail_port = 993;
+        if let Ok(login_param) = try_imap_usernames(context, param.clone(), imap).await {
+            return Ok(login_param);
+        }
 
-        if try_imap_one_param(context, &param, imap).await.is_ok() {
-            return Ok(());
+        // Try STARTTLS over port 143
+        param.server_flags &= !(DC_LP_IMAP_SOCKET_FLAGS as i32);
+        param.server_flags |= DC_LP_IMAP_SOCKET_STARTTLS as i32;
+        param.mail_port = 143;
+        try_imap_usernames(context, param, imap).await
+    } else if 0 == param.server_flags & DC_LP_SMTP_SOCKET_FLAGS as i32 {
+        // Try TLS over user-provided port.
+        param.server_flags &= !(DC_LP_IMAP_SOCKET_FLAGS as i32);
+        param.server_flags |= DC_LP_IMAP_SOCKET_SSL as i32;
+        if let Ok(login_param) = try_imap_usernames(context, param.clone(), imap).await {
+            return Ok(login_param);
         }
+
+        // Try STARTTLS over user-provided port.
+        param.server_flags &= !(DC_LP_IMAP_SOCKET_FLAGS as i32);
+        param.server_flags |= DC_LP_IMAP_SOCKET_STARTTLS as i32;
+        try_imap_usernames(context, param, imap).await
+    } else {
+        try_imap_usernames(context, param, imap).await
     }
+}
 
-    progress!(context, 660 + variation * 30);
+async fn try_imap_usernames(
+    context: &Context,
+    mut param: LoginParam,
+    imap: &mut Imap,
+) -> Result<LoginParam> {
+    if param.mail_user.is_empty() {
+        param.mail_user = param.addr.clone();
+        if let Ok(()) = try_imap_one_param(context, &param, imap).await {
+            return Ok(param);
+        }
 
-    if manually_set_param.mail_port == 0 {
-        param.mail_port = 143;
-        try_imap_one_param(context, param, imap).await
+        if let Some(at) = param.mail_user.find('@') {
+            param.mail_user = param.mail_user.split_at(at).0.to_string();
+        }
+        try_imap_one_param(context, &param, imap).await?;
```

Comment:
Because we are trying the same params twice if the username does not contain @: Obviously we need an Err to return so that I do not know any nice solution either.

But in case we decide to only push to a Vec here (@hpk42 's idea), we can do this in the if clause so that we try this params only once. (same below for smtp)

Reply:
Pushed a commit to fix it.